connector/failoverconnector/internal/state/pipeline_selector_test.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package state
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSelectPipeline(t *testing.T) {
+	constants := PSConstants{
+		RetryInterval: 50 * time.Millisecond,
+		RetryGap:      10 * time.Millisecond,
+		MaxRetries:    1000,
+	}
+	pS := NewPipelineSelector(5, constants, make(chan struct{}))
+
+	idx, ch := pS.SelectedPipeline()
+
+	require.Equal(t, idx, 0)
+	require.Equal(t, pS.ChannelIndex(ch), 0)
+}
+
+func TestHandlePipelineError(t *testing.T) {
+
+	done := make(chan struct{})
+
+	constants := PSConstants{
+		RetryInterval: 50 * time.Millisecond,
+		RetryGap:      10 * time.Millisecond,
+		MaxRetries:    1000,
+	}
+	pS := NewPipelineSelector(5, constants, done)
+
+	go pS.ListenToChannels(done)
+
+	idx, ch := pS.SelectedPipeline()
+
+	require.Equal(t, idx, 0)
+
+	ch <- false
+
+	require.Eventually(t, func() bool {
+		idx, _ = pS.SelectedPipeline()
+		return idx == 1
+	}, 3*time.Minute, 10*time.Millisecond)
+
+	close(done)
+
+}
+
+func TestCurrentPipelineWithRetry(t *testing.T) {
+
+	done := make(chan struct{})
+
+	constants := PSConstants{
+		RetryInterval: 50 * time.Millisecond,
+		RetryGap:      10 * time.Millisecond,
+		MaxRetries:    1000,
+	}
+	pS := NewPipelineSelector(5, constants, done)
+
+	go pS.ListenToChannels(done)
```

Comment:
Rather than close at the end, this should make sure we clean up after the test, regardless of the result.
```suggestion
	var wg sync.WaitGroup
	wg.Add(1)
	go pS.ListenToChannels(done)
	defer func() {
		close(done)
		wg.Wait()
	}()
```

Reply:
Good catch